Device power management support on cc13x2/cc26x2

vanti opened the pull request:
This PR adds device power management support for cc13x2/cc26x2 by leveraging TI's Power module. It updates the core device list and the following drivers to ensure they work correctly with the power management infrastructure in Zephyr:

- gpio
- uart
- spi
- i2c

~~This PR should be merged after #21465. I am putting this up early for review, so please ignore the first two commits for now.~~

vanti committed:
drivers: gpio: cc13xx_cc26xx: add power management support

Adding a dependency on the GPIO resource in the TI Power module, and
configuring the edge detection on pads as a wakeup source.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
drivers: serial: cc13xx_cc26xx: add power management support

Add a dependency on the UART resource in the TI Power module,
reconfigure the UART upon CPU exiting standby, and add support
for device PM.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
power: device: add core device list for cc13xx/cc26xx

Update the core device list to support device power management.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
drivers: spi: cc13xx_cc26xx: add power management support

Add a dependency on the SSI resource in the TI Power module, and add
support for device PM.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
drivers: i2c: cc13xx_cc26xx: add power management support

Add a dependency on the I2C resource in the TI Power module,
reconfigure the peripheral upon CPU exiting standby, and add
support for device PM.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
drivers: i2c: cc13xx_cc26xx: reset pin type when entering low power mode

It was observed that leaving the pin type configured for i2c for SDA
and SCL results in higher power consumption during standby. So we are
resetting the pin type when bringing the i2c into low power mode, and
setting it back to the correct type when exiting low power mode.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
power: policy: only bring devices to low-power when in sleep mode 2

Sleep mode 1 is supposed to be a low-latency sleep mode where devices
are left in active mode. Thus we should only bring devices to low-power
when in sleep mode 2 in sys_pm_policy_low_power_devices().

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
soc: cc13x2_cc26x2: set SYS_PM_STATE_LOCK when power management is enabled

When power management is enabled, SYS_PM_STATE_LOCK is necessary
in order to prevent the power policy from automatically entering
certain sleep states.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

galak reviewed (approved):
LGTM.

vanti commented:
Rebased now that #21465 has been merged

carlescufi commented:
@pabigot could you please review this PR?

pabigot commented on drivers/serial/uart_cc13xx_cc26xx.c:
```diff
@@ -53,6 +65,12 @@ static int uart_cc13xx_cc26xx_poll_in(struct device *dev, unsigned char *c)
 static void uart_cc13xx_cc26xx_poll_out(struct device *dev, unsigned char c)
 {
 	UARTCharPut(get_dev_conf(dev)->regs, c);
+	/*
```
Would it make sense to have this be conditional on power management?  Otherwise waiting for transmission to complete is a behavioral change that may impact other things the application would normally be doing after this function.

pabigot commented on drivers/serial/uart_cc13xx_cc26xx.c:
```diff
@@ -284,6 +351,119 @@ static void uart_cc13xx_cc26xx_isr(void *arg)
+			    DT_TI_CC13XX_CC26XX_UART_40001000_BASE_ADDRESS) {
+				Power_releaseDependency(
+					PowerCC26XX_PERIPH_UART0);
+				get_dev_data(dev)->pm_state = new_state;
```
Updating `pm_state` could be moved from branches to follow the conditional for clarity and to match the active state code.

pabigot commented on drivers/serial/uart_cc13xx_cc26xx.c:
```diff
@@ -313,6 +493,22 @@ static int uart_cc13xx_cc26xx_init_0(struct device *dev)
+	get_dev_data(dev)->rx_constrained = false;
+	get_dev_data(dev)->tx_constrained = false;
+
+	/* Set Power dependecies & constraints */
```
typo `dependecies`.  Five locations in three files.

pabigot commented on drivers/serial/uart_cc13xx_cc26xx.c:
```diff
@@ -313,6 +493,22 @@ static int uart_cc13xx_cc26xx_init_0(struct device *dev)
+#endif
+
+#ifdef CONFIG_SYS_POWER_MANAGEMENT
+	get_dev_data(dev)->rx_constrained = false;
```
Are these the right defaults when interrupt-driven UART is not enabled?  It looks like the constraints are only added as a result of enabling interrupts.  (IMO it'd be fine to require interrupt-driven UART if power management is enabled.)

pabigot replied:
My concern is that only that `SLEEP_2` (standby mode) is disabled if and only if the constrained variables are true (which only happens when interrupts are enabled, meaning the UART is active), and the comments suggest this is because UART won't work in standby mode.  So by inspection it seems that non-interrupt use of UART would fail if the system goes into standby.  (I'm concerned more about losing characters on reception because the UART is shut down; polled output will probably work because the system can't sleep during the spin-wait.)

At the moment the sleep states are default-enabled, though #21118 changes them to default-disabled.  (I expect that to get merged this development cycle.)

vanti replied:
Ahh, now I get it, thanks for the clarification - I was thinking of the poll-out. It is indeed a problem on the poll-in. The tricky part here is that even if CONFIG_UART_INTERRUPT_DRIVEN is thrown to enable interrupt mode, the non-interrupt mode functions are still available to be called and are not disabled. So users that are calling poll-in today would still suffer character lost.

Several alternatives I can think of:
1. Inside uart_cc13xx_cc26xx_poll_in, disable sleep state 2 if it is enabled, and issue a warning message to the log: "Using poll-in in uart driver automatically disables sleep state 2. Suggest to use interrupt-driven API if possible for better power savings". This would ensure code such as this would work regardless of of the value of SYS_PM_MIN_RESIDENCY_SLEEP_2:
```
        while (size) {
                u8_t c;
                res = uart_poll_in(tty->uart_dev, &c);

                /* Avoid 100% busy-polling, and yet try to process bursts
                 * of data without extra delays.
                 */
                if (res == -1) {
                        k_sleep(K_MSEC(1));
                }
        }
```

2. Change all the call sites in Zephyr to use interrupt-driven API instead of poll-in (e.g. tty.c), and make CONFIG_UART_INTERRUPT_DRIVEN disable the poll-in API. (will break existing code that relies on both at once)

3. Document in device-specific docs that poll-in may lose characters on this platform if sleep state 2 is enabled when PM is turned on (not very user-friendly for those who turn on PM without reading the docs)

Any thoughts on these? I am leaning towards #1 since it minimizes disruptions to other platforms, allows applications to at least work (albeit with sub-optimal PM), and removes the chance of someone scratching their head wondering why characters are getting lost.

pabigot replied:
I'm not sure.  I'm still trying to figure out the complexities of the UART driver.  I'm aware that poll-based UART is already a problem for power management (`uart_poll_out` at the API level doesn't guarantee the data is flushed before the call completes, so it can be lost generally even if you've patched things to make it work in this driver).  I was also thinking of cases where an enabled peripheral while sleeping would put something into the RX buffer where it could be picked up by a later `uart_poll_in()`.

The short-term solution I was thinking is that you can't put the UART into sleep mode when `CONFIG_UART_INTERRUPT_DRIVEN=n` but UART is enabled, so you would initialize the `?x_constrained` variables here to `true` and add the constraint at the same time.

The long-term solution is to come up with an API-level understanding of how power management affects UART, both poll and interrupt-driven.

vanti replied:
I tend to concur that the UART driver API would benefit from some adjustment to make power management easier. For example, if I compare this with the UART driver API in TI-RTOS, I see that they follow an open/close model (so that the device and its associated power domains can be powered only when in opened state at runtime, whereas in Zephyr we need to assume the device would be powered as soon as it is initialized and the CPU is in active state). They adhere to the concept of a UART transaction, so that the standby state in only locked for the duration of the transfer. As soon as the transaction is done the CPU can go into standby. So they do not have the issue we are seeing here, where the expectation is that the UART stays receptive to input even after the poll-in function has returned with its transaction essentially completed.

Afaik, `CONFIG_UART_INTERRUPT_DRIVEN=y` does not imply that all UART transfers would be interrupt-driven. Polling is still supported, so I don't think we can make the inference to set the constraints simply based on that flag.

For simplicity, I think I'll just document this issue in the short term, and we should aim for a UART API refinement in the longer term as you suggest.

pabigot commented on subsys/power/device.c:
```diff
@@ -29,6 +29,13 @@ static const char *const core_devices[] = {
+#elif defined(CONFIG_SOC_SERIES_CC13X2_CC26X2)
+#define MAX_PM_DEVICES	15
+#define MAX_DEV_NAME_LEN	16
+static const char core_devices[][MAX_DEV_NAME_LEN] = {
```
This should be `static const char *const core_devices[]` to match current master.

pabigot commented on subsys/power/policy/policy_residency_cc13x2_cc26x2.c:
```diff
@@ -140,3 +140,14 @@ enum power_states sys_pm_policy_next_state(s32_t ticks)
+
+__weak bool sys_pm_policy_low_power_devices(enum power_states pm_state)
+{
+	bool ret = false;
```
`return (pm_state == SYS_POWER_STATE_SLEEP_2);` ?

pabigot reviewed (commented):
A couple comments on serial also apply to the similar code in spi and i2c.

vanti reviewed (commented):
Thanks for the review @pabigot. I pushed some fixes and added some comments.

pabigot commented on drivers/spi/spi_cc13xx_cc26xx.c:
```diff
@@ -191,6 +205,73 @@ static int spi_cc13xx_cc26xx_release(struct device *dev,
+				DT_TI_CC13XX_CC26XX_SPI_40000000_BASE_ADDRESS) {
+				Power_releaseDependency(
+					PowerCC26XX_PERIPH_SSI0);
+				get_dev_data(dev)->pm_state = new_state;
```
This can also be moved out of the branches.

vanti committed:
boards: arm: add doc on power management and uart for cc13xx/cc26xx

Adding documentation to inform users that characters may be lost if
polling is used and a character arrives while the system is in standby
mode.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti commented:
Added documentation for poll-in issue.

nashif commented on subsys/power/device.c:
```diff
@@ -29,6 +29,12 @@ static const char *const core_devices[] = {
+static const char *const core_devices[] = {
+	"sys_clock",
+	"UART_0",
+};
```
we need a better way to deal with SOC related data, should not be in core implementation... But this is a story for another time.
